### PR TITLE
Test Windows

### PR DESCRIPTION
**Pull request by ocefpaf**
@krischer the first thing I did was to remove the [line](https://github.com/conda-forge/mtspec-feedstock/compare/master...ocefpaf:win?expand=1#diff-e178b687b10a71a3348107ae3154e44cL17) skipping the Windows build.

Note that the feedstock is not configured to build on Windows, so after I removed that I had to rerender it with `conda-smithy` (you can install it with `conda install -c conda-forge conda-smithy`). The command to rerender is: `conda smithy rerender`.

Note that the re-rendering created the `appveyor.yml` file automatically for us. (I am puzzled but the change in the README, the re-rendering should not do that. I must have an older version of `conda-smithy`.)

### CODE REVIEW

File: recipe/meta.yaml [hunk 1]
```diff
@@ -14,17 +14,17 @@ source:
...
   number: 0
   script: python setup.py install --single-version-externally-managed --record record.txt
-  skip: True  # [win]
 
 requirements:
   build:
     - python
     - setuptools
-    - gcc
+    - gcc  # [not win]
```
Comment (ocefpaf) [inline recipe/meta.yaml:27]: Now that we are building on Windows we need to limit `gcc` (and `libgfortran` below) to Linux and OS X.

File: recipe/meta.yaml [hunk 2]
```diff
@@ -14,17 +14,17 @@ source:
...
+    - mingwpy  # [win]
```
Comment (ocefpaf) [inline recipe/meta.yaml]: Let's try `mingwpy` first. But I am looking forward to test `msys2`.

**Comment by krischer** [comment]
Ok - it runs into the compiler problem I mentioned in the other issue - give me a couple minutes to create a patch.

**Comment by krischer** [comment]
Here you go (save as `win_compile.patch` next to the `meta.yaml`):

```
diff --git a/setup.py b/setup.py
index 00d0fcd..b77c1f7 100644
--- a/setup.py
+++ b/setup.py
@@ -84,7 +84,7 @@ else:
 def _compile(self, obj, src, ext, cc_args, extra_postargs, pp_opts):
     compiler_so = self.compiler_so
     if ext == ".f90":
-        if sys.platform == 'darwin' or sys.platform == 'linux2':
+        if True:
             compiler_so = ["gfortran"]
             cc_args = ["-O", "-fPIC", "-c", "-ffree-form"]
             if arch_flag:
```

This appears to work locally.

``` yaml
source:
  #[source information here]
  patches:
    - win_compile.patch 
```

**Comment by krischer** [comment]
Hmm...does not work and the error message is not very helpful 😞 I'll setup a windows VM later today except if you have some other ideas.

**Comment by ocefpaf** [comment]
> Hmm...does not work and the error message is not very helpful 😞 I'll setup a windows VM later today except if you have some other ideas.

Maybe `mingwpy` won't work here. I am testing `MSYS2` now.

@mingwandroid can you take a look if I got the `msys2` packages right?

**Comment by mingwandroid** [comment]
I strongly caution against ever using anything but `m2w64-toolchain` as a build requirement. You want all the other things expected as part of a Windows GCC-based toolchain (rc.exe, windres.exe, binutils etc etc). A fortran compiler on its own isn't much use.

Also, you have `m2w64-toolchain` as a run requirement, when I think you probably wanted `m2w64-gcc-libs` (or `m2w64-gcc-libgfortran` if you want to attempt to save a few MB - as I said earlier, I wouldn't bother).

**Comment by ocefpaf** [comment]
Thanks @mingwandroid! I am modifying to add `m2w64-toolchain` at both `build` and `run`, but should I add

``` yaml
    - m2w64-toolchain  # [win64]
    - m2w32-toolchain  # [win32]
```

or only `64` works on both?

**Comment by mingwandroid** [comment]
There is not such thing as `m2w32`, `m2w64` is an abbreviation of `msys2/mingw-w64`. `mingw-w64` is the 64-bit-capable fork of `mingw.org`, so you don't need any selectors at all here.

Though, I don't get why someone using this library needs a complete toolchain installed?

**Comment by ocefpaf** [comment]
> There is not such thing as `m2w32`, `m2w64` is an abbreviation of `msys2/mingw-w64`. `mingw-w64` is the 64-bit-capable fork of `mingw.org`, so you don't need any selectors at all here.

Just found out about that. Thanks!

> Though, I don't get why someone using this library needs a complete toolchain installed?

I will remove for the final build. Just testing it here. (I don't have a Windows machine to look into those, so I use AppVeyor as my VM :smile:)

**Comment by mingwandroid** [comment]
Ah OK. By the way, I've not got round to trying to build a Python module using `mingw-w64` yet. I expect there'll be a few things to fix up.

**Comment by mingwandroid** [comment]
Oh, you don't want `-fPIC` or `-fpic` on Windows, it always PIC, but it reminds you annoyingly about that fact with a warning.

**Comment by ocefpaf** [comment]
> Ah OK. By the way, I've not got round to trying to build a Python module using `mingw-w64` yet. I expect there'll be a few things to fix up.

Exiting new world :wink:  (But I will ping you soon for some R recipes :grimacing:)

> Oh, you don't want `-fPIC` or `-fpic` on Windows, it always PIC, but it reminds you annoyingly about that fact with a warning.

Got it. Thanks

**Comment by ocefpaf** [comment]
I got similar result with `mingwpy`. `x64` fails everywhere and `PY35` fails on `x86` too.

The only difference is that `msys2` was able to compiled the module on `PY26+x64` and the failure happens at test time while `minwpy` failed to compile.

``` shell
===== testing package: mtspec-0.3.1-py27_0 =====
import: u'mtspec'
===== mtspec-0.3.1-py27_0 OK =====
Traceback (most recent call last):
  File "C:\conda\envs\_test\lib\runpy.py", line 163, in _run_module_as_main
    mod_name, _Error)
  File "C:\conda\envs\_test\lib\runpy.py", line 102, in _get_module_details
    loader = get_loader(mod_name)
  File "C:\conda\envs\_test\lib\pkgutil.py", line 464, in get_loader
    return find_loader(fullname)
  File "C:\conda\envs\_test\lib\pkgutil.py", line 474, in find_loader
    for importer in iter_importers(fullname):
  File "C:\conda\envs\_test\lib\pkgutil.py", line 430, in iter_importers
    __import__(pkg)
  File "C:\conda\envs\_test\lib\site-packages\mtspec\__init__.py", line 18, in <module>
    from .multitaper import mtspec, sine_psd, dpss  # NOQA
  File "C:\conda\envs\_test\lib\site-packages\mtspec\multitaper.py", line 19, in <module>
    mtspeclib = _load_lib()
  File "C:\conda\envs\_test\lib\site-packages\mtspec\util.py", line 38, in _load_lib
    lib = C.CDLL(filename)
  File "C:\conda\envs\_test\lib\ctypes\__init__.py", line 362, in __init__
    self._handle = _dlopen(self._name, mode)
WindowsError: [Error 1114] A dynamic link library (DLL) initialization routine failed
TESTS FAILED: mtspec-0.3.1-py27_0
Command exited with code 1
```

@krischer we can skip the failing combinations and get some Windows binaries out there.

**Comment by krischer** [comment]
@ocefpaf Thanks a lot for your efforts.

> @krischer we can skip the failing combinations and get some Windows binaries out there.

I'll keep trying a bit - maybe we can get some more out of it :smile:

I really don't know a lot (nothing really) about compiling on windows, but a couple of thoughts:
- Python 3.5 is build with Visual Studio 2015 which apparently is not compatible with normal mingw compiled libraries (see e.g. here: https://github.com/numpy/numpy/pull/6875). I'm not sure this can work with normal  mingw (this is what `msys2` distributes, right)? We might have to use `mingwpy` whose explicit goal it is to produce compatible binaries.
- The `mingwpy` failure on 2.7 and 3.4 appears to be some linking issues which should be easy to fix. Maybe link against `libquadmath`?
- The error with `mingwpy` and 3.5 (seen in the following) appear to be quite common and thus can maybe also be fixed? This also happens with `msys2` on 64bit platforms.

```
  File "C:\conda\envs\_build\lib\distutils\cygwinccompiler.py", line 126, in __init__
    if self.ld_version >= "2.10.90":
TypeError: unorderable types: NoneType() >= str()
```
- The following error appears to be that `libc` or something is loaded in a wrong order or twice?  I could not really find a lot with google as this appears to be an extremely common windows error. I don't really know what to do about that to be honest.

```
WindowsError: [Error 1114] A dynamic link library (DLL) initialization routine failed
TESTS FAILED: mtspec-0.3.1-py27_0
Command exited with code 1
```

Can you give me push rights to your fork so I can play around a bit?

**Comment by ocefpaf** [comment]
> I'll keep trying a bit - maybe we can get some more out of it :smile:

Merging this is only to get what we have on the channel. I would open another PR right after it removing the constraints.

> We might have to use `mingwpy` whose explicit goal it is to produce compatible binaries.

In theory yes. But those fail with the current version of `mingwpy` and I am not sure there is a solution for that yet.

> The `mingwpy` failure on 2.7 and 3.4 appears to be some linking issues which should be easy to fix. Maybe link against `libquadmath`?

You mean `msys2` linking error? Sorry, I rebased and made it hard to follow...

> Can you give me push rights to your fork so I can play around a bit?

Sure. But we can just merge this, cash what we have, and open a new one to address those issues.
Or you can pick my commits and open a new PR to play with this.

**Comment by krischer** [comment]
> Merging this is only to get what we have on the channel. I would open another PR right after it removing the constraints.

Works for me! I think you still have to exclude py34 on window which builds but the tests don't run. I'm not sure why AppVeyor counts this as a successful build.

> You mean msys2 linking error? Sorry, I rebased and made it hard to follow...

No I meant `mingwpy` and 64 bit libs - but who knows - I had lots of tabs open and read the logs - maybe I confused stuff. I'll check once this is merged.

**Comment by ocefpaf** [comment]
>  I think you still have to exclude py34 on window which builds but the tests don't run.

Indeed. I think I read something about this somewhere, but pinging @msarahan here anyways so he is aware.

``` shell
    vs2010_runtime:          10.00.40219.1-2      defaults   
[conda] C:\conda\conda-bld\test-tmp_dir>call activate.bat C:\conda\envs\_test 
===== testing package: mtspec-0.3.1-py34_0 =====
import: 'mtspec'
Traceback (most recent call last):
  File "C:\conda\conda-bld\test-tmp_dir\run_test.py", line 26, in <module>
    import mtspec
  File "C:\conda\envs\_test\lib\site-packages\mtspec\__init__.py", line 18, in <module>
    from .multitaper import mtspec, sine_psd, dpss  # NOQA
  File "C:\conda\envs\_test\lib\site-packages\mtspec\multitaper.py", line 19, in <module>
    mtspeclib = _load_lib()
  File "C:\conda\envs\_test\lib\site-packages\mtspec\util.py", line 38, in _load_lib
    lib = C.CDLL(filename)
  File "C:\conda\envs\_test\lib\ctypes\__init__.py", line 351, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: [WinError 1114] A dynamic link library (DLL) initialization routine failed
Traceback (most recent call last):
  File "C:\conda\envs\_test\lib\runpy.py", line 151, in _run_module_as_main
    mod_name, mod_spec, code = _get_module_details(mod_name)
  File "C:\conda\envs\_test\lib\runpy.py", line 104, in _get_module_details
    spec = importlib.util.find_spec(mod_name)
  File "C:\conda\envs\_test\lib\importlib\util.py", line 86, in find_spec
    parent = __import__(parent_name, fromlist=['__path__'])
  File "C:\conda\envs\_test\lib\site-packages\mtspec\__init__.py", line 18, in <module>
    from .multitaper import mtspec, sine_psd, dpss  # NOQA
  File "C:\conda\envs\_test\lib\site-packages\mtspec\multitaper.py", line 19, in <module>
    mtspeclib = _load_lib()
  File "C:\conda\envs\_test\lib\site-packages\mtspec\util.py", line 38, in _load_lib
    lib = C.CDLL(filename)
  File "C:\conda\envs\_test\lib\ctypes\__init__.py", line 351, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: [WinError 1114] A dynamic link library (DLL) initialization routine failed
TEST END: mtspec-0.3.1-py34_0
# If you want to upload this package to anaconda.org later, type:
#
# $ anaconda upload C:\conda\conda-bld\win-32\mtspec-0.3.1-py34_0.tar.bz2
#
# To have conda build upload to anaconda.org automatically, use
# $ conda config --set anaconda_upload yes
Build success
```

[full log](https://ci.appveyor.com/project/conda-forge/mtspec-feedstock/build/1.0.10/job/f0wsk7lh1qs4cf8f)
